client/shared/src/api/util.ts

```diff
@@ -64,6 +64,65 @@ export const isSubscribable = (value: unknown): value is Subscribable<unknown> =
     hasProperty('subscribe')(value) &&
     typeof value.subscribe === 'function'
 
+/**
+ * Reports whether the value is an AsyncIterable
+ */
+export const isAsyncIterable = (value: unknown): value is AsyncIterable<unknown> =>
+    typeof value === 'object' &&
+    value !== null &&
+    // hasProperty doesn't work with Symbol.asyncIterator
+    (<O extends object>(object: O): object is O & { [Symbol.asyncIterator]: unknown } =>
+        hasProperty(Symbol.asyncIterator)(object))(value) &&
+    typeof value[Symbol.asyncIterator] === 'function'
+
+/**
+ * Convert an async iterable into an observable.
+ *
+ * @param iterable The source iterable.
+ */
+export const observableFromAsyncIterable = <T, R>(
+    iterable: AsyncGenerator<T, R> | AsyncIterable<T>
+): Observable<T | R> =>
+    new Observable((observer: Observer<T | R>) => {
+        const iterator = iterable[Symbol.asyncIterator]()
+        let unsubscribed = false
+        let iteratorDone = false
+        function next(): void {
+            iterator.next().then(
+                result => {
+                    if (unsubscribed) {
+                        return
+                    }
+                    if (result.done) {
+                        observer.next(result.value)
+                        iteratorDone = true
+                        observer.complete()
+                    } else {
+                        observer.next(result.value)
+                        return next()
+                    }
+                },
+                error => {
+                    observer.error(error)
+                }
+            )
+        }
+        next()
+        return () => {
+            unsubscribed = true
+            if (!iteratorDone && iterator.throw) {
+                iterator.throw(new AbortError()).catch(() => {
+                    // ignore
+                })
+            }
+        }
+    })
+
+export class AbortError extends Error {
+    public readonly name = 'AbortError'
+    public readonly message = 'Aborted'
```

Comment:
`message` would be nice to still allow to be overridden, and be kept non-enumerable as for normal errors.
```suggestion
    constructor(message: string = 'Aborted') {
        super(message)
    }
```

Reply:
Removed `AbortError`. I'll bring it back if we figure out a way to call `iterator.error()`